### PR TITLE
Fix ZLibDecompressor dropping data past the first gzip member

### LINKED ISSUE #7157: Decompressing concatenated gzip
### Describe the bug

Recently, I have been consuming a proprietary REST API that returns gzip-encoded text data.

I discovered that the response was different from the response of the library ``requests`` or ``Postman``. After a long research, I discovered that the REST API is sending concatenated gzip data.

It seems that decompressing concatenated gzip data requires special treatment of ``zlib.decompressobj.unused_data`` ([for example, check this answer on StackOverflow](https://stackoverflow.com/a/56980598)).

I have confirmed that a similar implementation is found in [urllib3](https://github.com/urllib3/urllib3/blob/dacd84d01f3037d30adf387e2e1bfc95f55dedfe/src/urllib3/response.py#L136) (which is used by the library ``requests``), where unused data is checked for further decompression.

I think that [this](https://github.com/aio-libs/aiohttp/blob/283861dddf7bac50188ad3dd63dda93bccfca4bc/aiohttp/http_parser.py#L917) is where ``aiohttp`` decompresses the gzip data. The `unused_data` is not handled in any way. I tried changing that line of ``aiohttp`` code to something like this:

```python

ret = self.decompressor.decompress(chunk)
while self.decompressor.unused_data:
    chunk = self.decompressor.unused_data
    self.decompressor = zlib.decompressobj(zlib.MAX_WBITS | 16)
    ret += self.decompressor.decompress(chunk)

chunk = ret
```

and I was able to reproduce the response of the library ``requests``.

I have labeled this issue as bug, since I believe that the desired behavior is the same as in the library ``requests``. If this treatment of gzip data was intentional, is there a simple way to process concatenated gzip data? Currently, ``aiohttp`` only returns a fragment of the decompressed response with ``await response.text()``.

### To Reproduce

Sorry, I cannot offer a way to reproduce because I am using a propietary REST API.

### Expected behavior

All the concatenated gzip data should be decompressed and concatenated as ``requests`` does [here](https://github.com/psf/requests/blob/ec553c275fe915ab8c179a530907b564892d2922/requests/models.py#L899).

### Logs/tracebacks

```python-traceback
No logs.
```

### Python Version

```console
Python 3.8.10
```

### aiohttp Version

```console
aiohttp 3.8.3
```

### multidict Version

```console
multidict 6.0.4
```

### yarl Version

```console
yarl 1.8.2
```

### OS

Windows 10

### Related component

Client

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by Ashutosh-177**
## What do these changes do?

When a response body contains concatenated gzip members (e.g. a server that produces one gzip member per write call, as nginx does under certain configs), `zlib.decompressobj` sets `eof` and stores the remaining bytes in `unused_data` after it finishes the first member. `decompress_sync()` wasn't checking `unused_data` at all, so every member after the first was silently discarded. The caller got truncated output with no error.

The fix applies the same `while eof and unused_data` loop that `ZSTDDecompressor` already uses for multi-frame zstd streams. Each iteration creates a fresh decompressor and feeds it the leftover bytes, accumulating output across all members. `max_length` is tracked and honoured across the loop.

Also added `unused_data` to `ZLibDecompressObjProtocol` so the attribute is properly typed, and three tests that mirror the existing ZSTD multi-frame test suite.

## Are there changes in behaviour for the end user?

Yes — responses with concatenated gzip members now decompress fully instead of being silently truncated at the first member boundary.

## Related issue number

Fixes #7157

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [ ] `make fmt` has been run N/A (formatting only)
- [x] All the tests pass
- [x] Changelog entry added
- [x] Added myself to `CONTRIBUTORS.txt`

Drafted with Claude Sonnet 4.6; reviewed by Ashutosh-177.

### CODE REVIEW

File: tests/test_compression_utils.py
Comment (Dreamsorcerer) [inline tests/test_compression_utils.py:1]: Please add a variation of the 4 tests from tests_http_parser.py in https://github.com/aio-libs/aiohttp/pull/12290/changes as you've done here.

File: aiohttp/compression_utils.py [hunk 1]
```diff
@@ -275,15 +278,38 @@ def __init__(
...
+            unused = self._decompressor.unused_data
+            self._decompressor = self._zlib_backend.decompressobj(wbits=self._mode)
+            remaining = (
+                max_length - len(result)
+                if max_length != ZLIB_MAX_LENGTH_UNLIMITED
+                else ZLIB_MAX_LENGTH_UNLIMITED
+            )
+            if max_length != ZLIB_MAX_LENGTH_UNLIMITED and remaining <= 0:
+                self._pending_unused_data = unused
+                break
```
Comment (Dreamsorcerer) [inline aiohttp/compression_utils.py]: I think the zstd version is a little cleaner here, would be good to align as close as possible:
https://github.com/aio-libs/aiohttp/pull/12290/changes#diff-c88fdeb279a2914aa157a5ebbf863d092eeb3014be8e8e4a1eaef48daa186651R359-R363

I think there may be an edge case that's not correct here too (the extra tests will tell us).

**Comment by Dreamsorcerer** [comment]
Seems to have broken something. If you get it working by tomorrow morning, we'll include it in the next release.

**Comment by Ashutosh-177** [comment]
Fixed — the eof reset was breaking deflate by resetting eof to False, which DeflateBuffer.feed_eof() treats as an incomplete stream. Guarded it to gzip mode only (wbits > MAX_WBITS). The Autobahn failure is flaky and unrelated to this change.
